src/plugins/roomslist/templates/roomslist.js

```diff
@@ -67,10 +67,44 @@ function tplRoomItem (el, room) {
         </div>`;
 }
 
+function tplRoomDomainGroup (el, domain, rooms) {
+    const i18n_title = __('Click to hide these rooms');
+    const collapsed = el.model.get('collapsed_domains');
+    const is_collapsed = collapsed.includes(domain);
+    return html`
+    <div class="muc-domain-group" data-domain="${domain}">
+        <a href="#" class="list-toggle muc-domain-group-toggle controlbox-padded" title="${i18n_title}" @click=${ev => el.toggleDomainList(ev, domain)}>
+            <converse-icon
+                class="fa ${ is_collapsed ? 'fa-caret-right' : 'fa-caret-down' }"
+                size="1em"
+                color="var(--muc-group-color)"></converse-icon>
+            ${domain}
+        </a>
+        <ul class="items-list muc-domain-group ${ is_collapsed ? 'collapsed' : '' }" data-domain="${domain}">
+            ${ rooms.map(room => tplRoomItem(el, room)) }
+        </ul>
+    </div>`;
+}
+
 export default (el) => {
     const { chatboxes, CHATROOMS_TYPE, CLOSED } = _converse;
+    const group_by_domain = api.settings.get('muc_grouped_by_domain');
     const rooms = chatboxes.filter(m => m.get('type') === CHATROOMS_TYPE);
     rooms.sort((a, b) => (a.getDisplayName().toLowerCase() <= b.getDisplayName().toLowerCase() ? -1 : 1));
+    // The rooms should stay sorted as they are iterated and added in order
+    const grouped_rooms = new Map();
+    if (group_by_domain) {
+        for (const room of rooms) {
+            const roomdomain = room.get('jid').split('@').at(-1).toLowerCase();
+            if (grouped_rooms.has(roomdomain)) {
+                grouped_rooms.get(roomdomain).push(room);
+            } else {
+                grouped_rooms.set(roomdomain, [room]);
+            }
+        }
+    }
+    const sorted_domains = Array.from(grouped_rooms.keys());
+    sorted_domains.sort();
```

Comment:
Would be a bit nicer if these lines 94 to 107 are put in a function that is lazily called on line 149.

It avoids unnecessarily creating a `Map` (i.e. when `group_by_domain` is `false`) and keeps the template function smaller.